gh-70870: Clarify dual usage of 'free variable' (#122545)
python/cpython: Doc/glossary.rst

      A variable defined in a class and intended to be modified only at
      class level (i.e., not in an instance of the class).

   closure variable
      A variable referenced from a :term:`nested scope` that is defined in an outer scope
      rather than being resolved at runtime from the globals or builtin namespaces.
      May be explicitly defined with the :keyword:`nonlocal` keyword to allow write access,

>       or implicitly defined if the variable is only read without being written to.

Reviewer: That's really dense. Should we add an example?

```suggestion
      A variable referenced from a :term:`nested scope` that is defined in an outer scope
      rather than being resolved at runtime from the globals or builtin namespaces.
      May be explicitly defined with the :keyword:`nonlocal` keyword to allow write access,
      or implicitly defined if the variable is only read without being written to.
      
      For example, in the ``mid`` function in the following code,
      both ``a`` an ``b`` are *closure variables*,
      but only ``a`` is a :term:`free variable`::

         def outer():
             a = 1
             b = 2
             def mid():
                 print(a)
                 def inner():
                     print(b)

      
```
and add ```See :term:`closure variable` for an example``` to the other one?

Author: Both `a` and `b` are closure *and* free variables in `mid` (due to lexical scoping, every free variable in `inner` is also a free variable in `mid`, even if `mid` doesn't otherwise reference the name).

In the example, `print` is a free variable without being a closure variable, though.

```suggestion
      A :term:`free variable` referenced from a :term:`nested scope` that is defined in an outer
      scope rather than being resolved at runtime from the globals or builtin namespaces.
      May be explicitly defined with the :keyword:`nonlocal` keyword to allow write access,
      or implicitly defined if the variable is only read without being written to.
      
      For example, in the ``inner`` function in the following code, both ``x`` and ``print`` are
      :term:`free variables <free variable>`, but only ``x`` is a *closure variable*::
      
          def outer():
              x = 0
              def inner():
                  nonlocal x
                  x += 1
                  print(x)
              return inner
```

      A variable defined in a class and intended to be modified only at
      class level (i.e., not in an instance of the class).

   closure variable
      A :term:`free variable` referenced from a :term:`nested scope` that is defined in an outer
      scope rather than being resolved at runtime from the globals or builtin namespaces.
      May be explicitly defined with the :keyword:`nonlocal` keyword to allow write access,

>       or implicitly defined if the variable is only read without being written to.

Reviewer 2: ```suggestion
      or implicitly defined if the variable is only being read.
```

      For example, in the ``inner`` function in the following code, both ``x`` and ``print`` are
      :term:`free variables <free variable>`, but only ``x`` is a *closure variable*::

          def outer():
              x = 0
              def inner():
                  nonlocal x
                  x += 1
                  print(x)
              return inner

>       Due to the name of the :attr:`codeobject.co_freevars` attribute, sometimes instead
>       referred to as a :term:`free variable`.

Reviewer 2: ```suggestion
      Closure variables are sometimes
      referred to as a :term:`free variable` due to the name of the :attr:`codeobject.co_freevars` attribute.
```

Author: I realised this wasn't quite saying what I meant. Calling a closure variable a free variable is fine, since closure variables are a subset of free variables. The misleading phrasing is when people say "free variables" when only referring to a function's "closure variables".

```suggestion
      Due to the :attr:`codeobject.co_freevars` attribute (which, despite its name, only
      includes the names of closure variables rather than listing all referenced free
      variables), the more general :term:`free variable` term is sometimes used even
      when the intended meaning is to refer specifically to closure variables.
```

Author: I think this phrasing is still awkward, but it's at least explicit as to what I meant. If anyone comes up with some better wording later, we'll at least have a framework to include it now.

Reviewer 2: Love the explicit wording. I think that works well @ncoghlan.

      the :term:`global interpreter lock` which allows only one thread to
      execute Python bytecode at a time.  See :pep:`703`.

   free variable
      Formally, as defined in the :ref:`language execution model <bind_names>`, a free
      variable is any variable used in a namespace which is not a local variable in that

>       namespace.

Author: ```suggestion
      namespace. See :term:`closure variable` for an example.
```
